torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -3797,6 +3798,90 @@ def test_ddp_comm_hook_logging(self):
+            BACKEND != "nccl" and BACKEND != "gloo",
+            "Only Nccl & Gloo backend support DistributedDataParallel",
+        )
+        @skip_if_lt_x_gpu(2)
+        def test_ddp_hook_with_optimizer_parity(self):
+            rank = self.rank
+            torch.cuda.set_device(rank)
+            model = LargeNet()
+            sgd_lr = 1e-2
+            ddp_model_with_optimizer_hook = torch.nn.parallel.DistributedDataParallel(
+                copy.deepcopy(model).cuda(),
+                device_ids=[self.rank],
+            )
+            ddp_model_with_allreduce_hook = torch.nn.parallel.DistributedDataParallel(
+                copy.deepcopy(model).cuda(),
+                device_ids=[self.rank],
+            )
+            sgd_allreduce_hook = torch.optim.SGD(
+                ddp_model_with_allreduce_hook.parameters(),
+                lr=sgd_lr
+             )
+            ddp_model_with_no_hook = torch.nn.parallel.DistributedDataParallel(
+                copy.deepcopy(model).cuda(),
+                device_ids=[self.rank],
+            )
+            sgd_no_hook = torch.optim.SGD(
+                ddp_model_with_no_hook.parameters(),
+                lr=sgd_lr
+            )
+            allreduce_hook = default.allreduce_hook
+            opt_hook_state = default.OptimizerHookState(
+                _FunctionalSGD,
+                sgd_lr,
+            )
+            ddp_model_with_optimizer_hook.register_comm_hook(
+                None,
+                default.hook_then_optimizer(allreduce_hook, opt_hook_state),
+            )
+            ddp_model_with_allreduce_hook.register_comm_hook(
+                None,
+                default.allreduce_hook,
+            )
+
+            inp = torch.randn(1, 1000).cuda()
+            # Verify params initially
+            for hook_param, allreduce_param, no_hook_param in zip(
+                ddp_model_with_optimizer_hook.parameters(),
+                ddp_model_with_allreduce_hook.parameters(),
+                ddp_model_with_no_hook.parameters()
+            ):
+                self.assertEqual(hook_param, allreduce_param)
+                self.assertEqual(hook_param, no_hook_param)
+
+            old_hook_params = copy.deepcopy(
+                list(ddp_model_with_optimizer_hook.parameters())
+            )
+            for i in range(6):
+                out = ddp_model_with_optimizer_hook(inp)
+                loss = out.sum()
+                loss.backward()
+                out = ddp_model_with_allreduce_hook(inp)
+                loss = out.sum()
+                loss.backward()
+                sgd_allreduce_hook.step()
+                out = ddp_model_with_no_hook(inp)
+                loss = out.sum()
+                loss.backward()
+                sgd_no_hook.step()
+                for hook_param, allreduce_param, no_hook_param in zip(
+                    ddp_model_with_optimizer_hook.parameters(),
+                    ddp_model_with_allreduce_hook.parameters(),
```

Comment:
if `ddp_model_with_optimizer_hook` already gives the same results as the no-hook case, any reason we need to further compare with `ddp_model_with_allreduce_hook`?

Reply:
@mrshenli Good point, and we already do test hook parity vs local for allreduce in separate tests (`test_ddp_hook_parity`). I'll update the test to test between model with hook + optimizer versus regular model.